uast/uast-encoding-v2.md

```diff
@@ -0,0 +1,249 @@
+# UAST v2 binary encoding
+
+UAST binary encoding uses [Protocol Buffers](https://developers.google.com/protocol-buffers/)
+to encode individual messages, but requires additional logic to reconstruct the UAST.
+
+The design document can be found [here](https://docs.google.com/document/d/1iDOuWLnTckBdSscUMWGl2hBqAAqRn8jdRabXZSdr2aw/edit?usp=sharing)
+and the reference implementation in Go can be found [in SDK](https://github.com/bblfsh/sdk/blob/b44cc02214bf0d9c55dc3b830f423a6df6709ca9/uast/nodes/nodesproto/nodesproto.go).
+
+On a high level, a binary UAST uses [UAST v2 data model](./representation-v2.md).
+It extends this model by adding an ephemeral node ID field used to link UAST nodes.
+
+The file consist of two sections:
+
+* [Header](#header), containing "magic", format version and other options.
+* Flat [list of nodes](#nodes) linked by node IDs.
+
+The list contains zero or more graph nodes addressed by ID. Binary format version `1` must
+only store tree structures, but the format was made compatible with graph structure for future
+extensions.
+
+The ID of the root tree node is stored in the [Header](#header) message. Optionally an ID
+of a second tree root can be provided to store file metadata. Implementations are free to ignore it.
```

Comment:
This was intended for the future if we decide to return some additional info with the AST (e.g. detected language, file name, etc). Right now it's unused. Assuming graphs, we may even drop it in the next spec revision.

Reply:
Makes sense. Perhaps for this summary, then, we could simplify this to just the essential part:

```
The ID of the root tree node is located in the [Header](#header) message.
```

It doesn't preclude adding it back later, but the reader doesn't really need to do anything about it since it's unused and optional.